R-package/R/create-release-files.R

```diff
@@ -0,0 +1,411 @@
+              validated_file <- TRUE
+            } else{
+              logger::log_warn(paste0(single_part,
+                                      " does not have a matching sheet."))
+              errors <- TRUE
+            }
+            
+          } else {
+            logger::log_warn(paste0(partID,
+                                    " is not found in parts sheet."))
+            errors <- TRUE
+          }
+          
+        }
+      } else{
+        logger::log_warn(paste0(partID,
+                                " has an unrecognized fileType of ",
+                                file_type))
+        errors <- TRUE
+      }
+      
+      # Append valid files list
+      if (validated_file) {
+        files_to_extract[[fileID]] <- list(
+          file_name = file_name,
+          file_type = file_type,
+          partID = partID,
+          add_headers = add_headers,
+          destination = destination,
+          osf_location = osf_location,
+          github_location = github_location
+        )
+      }
+    }
+    # Will move with development
+    # Will become stop once function development is finished
+    if (errors != "") {
+      
+      warning("Errors were detected further building cannot continue please check the log for additional info")
+    }
+    return(files_to_extract)
+  }
+
+#' Create Files
+#'
+#' Function responsible for creating the release files based on output from validate_files_sheet
+#'
+#' @param files_to_extract List output from validate_files_sheet
+#' @param dictionary openxlsx environment object storing the dictionary
+create_files <-
+  function(files_to_extract,
+           dictionary) {
+    # Loop over files to extract based on fileID
+    for (fileID in names(files_to_extract)) {
+      current_file_info <- files_to_extract[[fileID]]
+      
+      # Create a write_directory based on destination and saving location
+      write_dir <- ""
+      if (current_file_info$destination == "github") {
+        write_dir <- file.path(
+          getwd(),
+          odm_dictionary$tmp_dictionary_directory,
+          "github",
+          current_file_info$github_location
+        )
+        dir.create(write_dir,
+                   showWarnings = FALSE,
+                   recursive = TRUE)
+        
+      } else if (current_file_info$destination == "osf") {
+        write_dir <- file.path(
+          getwd(),
+          odm_dictionary$tmp_dictionary_directory,
+          "osf",
+          current_file_info$osf_location
+        )
+        dir.create(write_dir,
+                   showWarnings = FALSE,
+                   recursive = TRUE)
+      }
+      
+      
+      if (current_file_info$file_type == "excel") {
+        # Use parts as names of sheets to extract
+        sheets_to_read <- current_file_info$partID
+        
+        # Create vector to track recorded sheets
+        sheet_was_copied <- rep(FALSE, length(sheets_to_read))
+        names(sheet_was_copied) <- sheets_to_read
```

Comment:
Do we need this? Can't we just remove each sheet as its "copied" from `sheets_to_read`?

Reply:
Yes it is no longer needed. It's left over from when I was looping over sheets_to_read and copying sheets over when possible. And mutating a list you are looping over is extremely unstable and does not work. So I used this to keep track of sheets that were successfully copied. However since formatting was not kept when copying it became better to copy full workbook and strip off anything not needed. So now it is no longer needed and I will remove it.

Reply:
Yeah, you should never mutate a list you are iterating over, i.e. deleting or inserting new elements. Updating the elements in a list should be ok.

I believe in C#, the compiler actually throws an error if it detects that you're doing this.